File: src/main/kotlin/dev/jsinco/recipes/configuration/RecipesTranslator.java
package dev.jsinco.recipes.configuration;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableMap;
import net.kyori.adventure.key.Key;
import net.kyori.adventure.text.minimessage.translation.MiniMessageTranslator;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.Reader;
import java.io.Writer;
import java.net.URISyntaxException;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.DirectoryStream;
import java.nio.file.FileSystem;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Enumeration;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Properties;

//TODO make locale configurable /// placeholder

/**
 * Copy-pasted from TBP, can't bother to migrate to kotlin
 */
public class RecipesTranslator extends MiniMessageTranslator {

    private Map<Locale, Properties> translations;
    private final File localeDirectory;

    public RecipesTranslator(File localeDirectory) {
        this.localeDirectory = localeDirectory;
    }

    public void reload() {
        syncLangFiles();
        loadLangFiles();
    }

    private void syncLangFiles() {
        if (!localeDirectory.exists() && !localeDirectory.mkdirs()) { // shouldn't happen
            throw new IllegalStateException("Failed to create locale directory at " + localeDirectory.getAbsolutePath());
        }

        try {
            Enumeration<URL> resources = getClass().getClassLoader().getResources("locale");
            while (resources.hasMoreElements()) {
                URL url = resources.nextElement();

                try (FileSystem fs = "jar".equals(url.getProtocol()) ? FileSystems.newFileSystem(url.toURI(), Collections.emptyMap()) : null) {

                    Path internalLocaleDir = fs == null ? Paths.get(url.toURI()) : fs.getPath("locale");
                    try (DirectoryStream<Path> stream = Files.newDirectoryStream(internalLocaleDir, "*.properties")) {
                        for (Path path : stream) {
                            mergeAndStoreProperties(path);
                        }
                    }
                }
            }
        } catch (IOException | URISyntaxException e) {
            throw new RuntimeException("Failed to sync language files", e);
        }
        // special thanks to StackOverflow and other useful sites lol
    }

    private void mergeAndStoreProperties(Path internalFile) throws IOException {
        String fileName = internalFile.getFileName().toString();
        File externalFile = new File(localeDirectory, fileName);

        Properties internalProps = new Properties();
        try (Reader reader = Files.newBufferedReader(internalFile, StandardCharsets.UTF_8)) {
            internalProps.load(reader);
        }

        Properties merged = new Properties();
        if (externalFile.exists()) {

            Properties externalProps = new Properties();
            try (Reader reader = new InputStreamReader(new FileInputStream(externalFile), StandardCharsets.UTF_8)) {
                externalProps.load(reader);
            }

            merged.putAll(externalProps);
            for (String key : internalProps.stringPropertyNames()) {
                merged.putIfAbsent(key, internalProps.getProperty(key));
            }

        } else {
            merged.putAll(internalProps);
            if (!externalFile.createNewFile()) {
                throw new IOException("Could not create file: " + externalFile);
            }
        }
        try (Writer writer = new OutputStreamWriter(new FileOutputStream(externalFile), StandardCharsets.UTF_8)) {
            storeWithoutComments(merged, writer);
        }
    }

    private void storeWithoutComments(Properties props, Writer writer) throws IOException {
        List<String> keys = new ArrayList<>(props.stringPropertyNames());
        Collections.sort(keys);

        for (String key : keys) {
            writer.write(key + "=" + props.getProperty(key) + "\n");
        }
    }

    private void loadLangFiles() {
        if (!localeDirectory.isDirectory()) {
            throw new IllegalArgumentException("Locale directory is not a directory!");
        }
        ImmutableMap.Builder<Locale, Properties> translationsBuilder = new ImmutableMap.Builder<>();
        for (File translationFile : localeDirectory.listFiles(file -> file.getName().endsWith(".properties"))) {
            try (InputStream inputStream = new FileInputStream(translationFile)) {
                Properties translation = new Properties();
                translation.load(new InputStreamReader(inputStream, StandardCharsets.UTF_8));
                Locale locale = Locale.forLanguageTag(translationFile.getName().replaceAll(".properties$", ""));
                if (locale != null) {
                    translationsBuilder.put(locale, translation);
                }
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
        this.translations = translationsBuilder.build();
        Preconditions.checkArgument(translations.containsKey(Locale.ENGLISH), "Unknown translation: " + Locale.ENGLISH);
    }

    @Override
    public @NotNull Key name() {
        return Key.key("recipes:global_translator");
    }

    @Override
    public @Nullable String getMiniMessageString(@NotNull String key, @NotNull Locale locale) {
        Properties translations = this.translations.get(Locale.ENGLISH);
        Preconditions.checkState(translations != null, "Should have found a translation!");
        return translations.getProperty(key);
    }
}

File: src/main/kotlin/dev/jsinco/recipes/configuration/RecipesTranslator.kt
package dev.jsinco.recipes.configuration

import com.google.common.base.Preconditions
import com.google.common.collect.ImmutableMap
import net.kyori.adventure.key.Key
import net.kyori.adventure.text.minimessage.translation.MiniMessageTranslator
import java.io.*
import java.net.URISyntaxException
import java.nio.charset.StandardCharsets
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.*

class RecipesTranslator(private val localeDirectory: File) : MiniMessageTranslator() {

    private var translations: Map<Locale, Properties>

    init {
        syncLangFiles()
        translations = loadLangFiles()
    }

    fun reload() {
        syncLangFiles()
        translations = loadLangFiles()
    }

    private fun syncLangFiles() {
        check(!(!localeDirectory.exists() && !localeDirectory.mkdirs())) { "Failed to create locale directory at " + localeDirectory.absolutePath }

        try {
            val resources = javaClass.getClassLoader().getResources("locale")
            while (resources.hasMoreElements()) {
                val url = resources.nextElement()

                (if ("jar" == url.protocol) FileSystems.newFileSystem(
                    url.toURI(),
                    mutableMapOf<String?, Any?>()
                ) else null).use { fs ->
                    val internalLocaleDir = Paths.get(url.toURI())
                    Files.newDirectoryStream(internalLocaleDir, "*.properties").use { stream ->
                        for (path in stream) {
                            mergeAndStoreProperties(path)
                        }
                    }
                }
            }
        } catch (e: IOException) {
            throw RuntimeException("Failed to sync language files", e)
        } catch (e: URISyntaxException) {
            throw RuntimeException("Failed to sync language files", e)
        }
        // special thanks to StackOverflow and other useful sites lol
    }

    @Throws(IOException::class)
    private fun mergeAndStoreProperties(internalFile: Path) {
        val fileName = internalFile.fileName.toString()
        val externalFile = File(localeDirectory, fileName)

        val internalProps = Properties()
        Files.newBufferedReader(internalFile, StandardCharsets.UTF_8).use { reader ->
            internalProps.load(reader)
        }
        val merged = Properties()
        if (externalFile.exists()) {
            val externalProps = Properties()
            InputStreamReader(FileInputStream(externalFile), StandardCharsets.UTF_8).use { reader ->
                externalProps.load(reader)
            }
            merged.putAll(externalProps)
            for (key in internalProps.stringPropertyNames()) {
                merged.putIfAbsent(key, internalProps.getProperty(key))
            }
        } else {
            merged.putAll(internalProps)
            if (!externalFile.createNewFile()) {
                throw IOException("Could not create file: $externalFile")
            }
        }
        OutputStreamWriter(FileOutputStream(externalFile), StandardCharsets.UTF_8).use { writer ->
            storeWithoutComments(merged, writer)
        }
    }

    @Throws(IOException::class)
    private fun storeWithoutComments(props: Properties, writer: Writer) {
        val keys: MutableList<String> = ArrayList<String>(props.stringPropertyNames())
        keys.sort()

        for (key in keys) {
            writer.write(key + "=" + props.getProperty(key) + "\n")
        }
    }

    private fun loadLangFiles() : Map<Locale, Properties> {
        require(localeDirectory.isDirectory()) { "Locale directory is not a directory!" }
        val translationsBuilder = ImmutableMap.Builder<Locale, Properties>()
        for (translationFile in localeDirectory.listFiles { file: File? ->
            file?.getName()?.endsWith(".properties") ?: false
        }) {
            try {
                FileInputStream(translationFile).use { inputStream ->
                    val translation = Properties()
                    translation.load(InputStreamReader(inputStream, StandardCharsets.UTF_8))
                    val locale = Locale.forLanguageTag(translationFile.getName().replace(".properties$".toRegex(), ""))
                    if (locale != null) {
                        translationsBuilder.put(locale, translation)
                    }
                }
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
        val output = translationsBuilder.build()
        Preconditions.checkArgument(
            output.containsKey(Locale.ENGLISH),
            "Unknown translation: " + Locale.ENGLISH
        )
        return output
    }

    override fun name(): Key {
        return Key.key("recipes:global_translator")
    }

    public override fun getMiniMessageString(key: String, locale: Locale): String? {
        val translation: Properties? = this.translations[Locale.ENGLISH]
        Preconditions.checkState(translation != null, "Should have found a translation!")
        return translation!!.getProperty(key)
    }
}